README.md
# i3-instant-layout


Automatic 'list based' layouts for the [i3](https://i3wm.org) window manager

## Animated summary
![Demo of i3-instant-layout](https://github.com/TyberiusPrime/i3-instant-layout/raw/master/docs/_static/i3-instant-layout_demo.gif "i3-instant-layout demo")


## Description

This python program drags i3 into the 'managed layouts
tiling window manager world' kicking and screaming.

What it does is apply a window layout to your current workspace,
like this one:

    -------------
    |     |  2  |
    |     |-----|
    |  1  |  3  |
    |     |-----|
    |     |  4  |
    -------------

The big advantage here  is that it needs no 'swallow' definitions whatsoever,
it's 'instant' - just add milk, eh, press the button.

## Get started

To get started, install with `pip install i3-instant-layout`, or if you prefer, [pipx](https://github.com/pipxproject/pipx)
and add this to your i3 config: 
`bindsym $mod+Escape exec "i3-instant-layout --list | rofi -dmenu -i | i3-instant-layout -` (or use the interactive menu of your choice).


## Further information

Call `i3-instant-layout --help` for full details, or 
`i3-instant-layout --desc` for the full list of supported layouts (or see below).


## Helpful tips

### How to sort windows
To get the windows in the right order for your layout of choice,
first enable the vStack or hStack layout, sort them,
and the proceed to your layout of choice.

### Border styles

i3-instant-layout must unmap/map the windows (ie. hide them temporarily) for i3 
to place them at the rigth location.
Unfortunatly that appears to consume the border style.
Work around this with a line like this in your i3 config:
```
for_window [class="^.*"] border pixel 1
```




## Available layouts

Layout: vStack

Aliases: ['1col', '1c']

One column / a vertical stack.

	---------
	|   1   |
	---------
	|   2   |
	---------
	|   3   |
	---------


--------------------------------------------------------------------------------

Layout: hStack

Aliases: ['1row', '1r']

One row / a horizontal stack

	-------------
	|   |   |   |
	| 1 | 2 | 3 |
	|   |   |   |
	-------------


--------------------------------------------------------------------------------

Layout: v2Stack

Aliases: ['2col', '2c', '2v']

Two columns of stacks
	
	-------------
	|  1  |  4  |
	-------------
	|  2  |  5  |
	-------------
	|  3  |  6  |
	-------------


--------------------------------------------------------------------------------

Layout: h2Stack

Aliases: ['2row', '2r', '2h']

Two rows of stacks
	
	-------------------
	|  1  |  2  |  3  |
	-------------------
	|  4  |  5  |  6  |
	-------------------


--------------------------------------------------------------------------------

Layout: v3Stack

Aliases: ['3col', '3c', '3v']

Three columns of stacks

	-------------------
	|  1  |  3  |  5  |
	-------------------
	|  2  |  4  |  6  |
	-------------------


--------------------------------------------------------------------------------

Layout: h3Stack

Aliases: ['3row', '3r', '3h']

Three rows of stacks

	-------------------
	|  1  |  2  |  3  |
	-------------------
	|  4  |  5  |  6  |
	-------------------
	|  7  |  8  |  9  |
	-------------------


--------------------------------------------------------------------------------

Layout: max

Aliases: ['maxTabbed']

One large container, in tabbed mode.

	---------------
	|             |
	|   1,2,3,4,  |
	|             |
	---------------


--------------------------------------------------------------------------------

Layout: mainLeft

Aliases: ['ml', 'mv', 'MonadTall']

One large window to the left at 50%,
all others stacked to the right vertically.

	-------------
	|     |  2  |
	|     |-----|
	|  1  |  3  |
	|     |-----|
	|     |  4  |
	-------------


--------------------------------------------------------------------------------

Layout: mainRight

Aliases: ['mr', 'vm', 'MonadTallFlip']

One large window to the right at 50%,
all others stacked to the right vertically.

	-------------
	|  2  |     |
	|-----|     |
	|  3  |  1  |
	|-----|     |
	|  4  |     |
	-------------


--------------------------------------------------------------------------------

Layout: MainMainVStack

Aliases: ['mmv']

Two large windows to the left at 30%,
all others stacked to the right vertically.

	-------------------
	|     |     |  3  |
	|     |     |-----|
	|  1  |  2  |  4  |
	|     |     |-----|
	|     |     |  5  |
	-------------------


--------------------------------------------------------------------------------

Layout: MainVStackMain

Aliases: ['mvm']

Two large windows at 30% to the left and right,
a vstack in the center

	-------------------
	|     |  3  |     |
	|     |-----|     |
	|  1  |  4  |  2  |
	|     |-----|     |
	|     |  5  |     |
	-------------------


--------------------------------------------------------------------------------

Layout: matrix

Aliases: []

Place windows in a n * n matrix.

The matrix will place swallow-markers
if you have less than n*n windows.

N is math.ceil(math.sqrt(window_count))


--------------------------------------------------------------------------------

Layout: VerticalTileTop

Aliases: ['vtt']

Large master area (66%) on top,
horizontal stacking below


--------------------------------------------------------------------------------

Layout: VerticalTileBottom

Aliases: ['vtb']

Large master area (66%) on bottom,
horizontal stacking above


--------------------------------------------------------------------------------

Layout: NestedRight

Aliases: ['nr']

Nested layout, starting with a full left half.


	-------------------------
	|           |           |
	|           |     2     |
	|           |           |
	|     1     |-----------|
	|           |     |  4  |
	|           |  3  |-----|
	|           |     |5 | 6|
	-------------------------


--------------------------------------------------------------------------------

Layout: SmartNestedRight

Aliases: ['snr']

Nested layout, starting with a full left half,
but never going below 1/16th of the size.

	2 windows
	-------------------------
	|           |           |
	|           |           |
	|           |           |
	|     1     |     2     |
	|           |           |
	|           |           |
	|           |           |
	-------------------------

	5 windows
	-------------------------
	|           |           |
	|           |     2     |
	|           |           |
	|     1     |-----------|
	|           |     |  4  |
	|           |  3  |-----|
	|           |     |  5  |
	-------------------------

	6 windows
	-------------------------
	|           |           |
	|           |     2     |
	|           |           |
	|     1     |-----------|
	|           |  3  |  4  |
	|           |-----|-----|
	|           |  5  |  6  |
	-------------------------

	7 windows
	-------------------------
	|           |     |     |
	|           |  2  |  3  |
	|           |     |     |
	|     1     |-----------|
	|           |  4  |  5  |
	|           |-----|-----|
	|           |  6  |  7  |
	-------------------------


	15 windows
	-------------------------
	|     |  2  |  4  |  6  |
	|  1  |-----|-----|-----|
	|     |  3  |  5  |  7  |
	|-----------|-----------|
	|  8  |  A  |  C  |  E  |
	|-----|-----|-----|-----|
	|  9  |  B  |  D  |  F  |
	-------------------------

Falls back to matrix layout above 16 windows.

--------------------------------------------------------------------------------

Layout: mainCenter

Aliases: ['mc', 'vmv']
One large window in the midle at 50%,
all others stacked to the left/right vertically.

	-------------------
	|  2  |     |  5  |
	|-----|     |-----|
	|  3  |  1  |  6  |
	|-----|     |-----|
	|  4  |     |  7  |
	-------------------

--------------------------------------------------------------------------------



CHANGELOG.md
# Changelog

## Version 0.15

* added layout MainCenter, vmv

## Version 0.14

initial public release



tests/conftest.py
# -*- coding: utf-8 -*-
"""
    Dummy conftest.py for i3_autolayout.

    If you don't know what this is for, just leave it empty.
    Read more about conftest.py under:
    https://pytest.org/latest/plugins.html
"""

# import pytest


src/i3_instant_layout/main.py
import asyncio
import i3ipc
import datetime
import json
import math
import subprocess
import sys
import tempfile
from pathlib import Path
from . import layouts


counter_file = Path("~/.local/share/i3-instant-layout/counter.json").expanduser()
counter_file.parent.mkdir(exist_ok=True, parents=True)


def append_layout(layout_dict, window_count):
    """Apply a layout from this layout class"""
    tf = tempfile.NamedTemporaryFile(suffix=".json")
    tf.write(json.dumps(layout_dict, indent=4).encode("utf-8"))
    tf.flush()
    cmd = ["i3-msg", "append_layout", str(Path(tf.name).absolute())]
    subprocess.check_call(cmd, stdout=subprocess.PIPE)
    tf.close()


def nuke_swallow_windows():
    """Remove swallow windows before changing layout"""
    to_nuke = set()

    def walk_tree(con):
        if con.ipc_data.get("swallows", False):
            to_nuke.add(con.ipc_data["window"])
        for d in con.descendants():
            walk_tree(d)

    i3 = i3ipc.Connection()
    tree = i3.get_tree().find_focused().workspace()
    walk_tree(tree)
    for window_id in to_nuke:
        subprocess.check_call(["xdotool", "windowclose", str(window_id)])


def get_window_ids():
    """use xprop to list windows on current screen.

    Couldn't find out how to get the right ids from i3ipc.
    id is con_id, but we need x11 id

    Sorry, this probably means this won't work on wayland & sway.
    """

    desktop = subprocess.check_output(
        ["xprop", "-notype", "-root", "_NET_CURRENT_DESKTOP"]
    ).decode("utf-8", errors="replace")
    desktop = desktop[desktop.rfind("=") + 2 :].strip()
    res = subprocess.check_output(
        [
            "xdotool",
            "search",
            "--all",
            "--onlyvisible",
            "--desktop",
            desktop,
            "--class",
            "",
        ]
    ).decode("utf-8", errors="replace")
    return res.strip().split("\n")


def get_active_window():
    return (
        subprocess.check_output(["xdotool", "getactivewindow"]).decode("utf-8").strip()
    )


def focus_window(id):
    return subprocess.check_call(
        ["i3-msg", f'[id="{id}"]', "focus"], stdout=subprocess.PIPE
    )
    # return subprocess.check_call(['xdotool','windowraise', id])


def apply_layout(layout, dry_run=False):
    """Actually turn this workspace into this layout"""
    active = get_active_window()
    windows = get_window_ids()
    window_count = len(windows)
    # we unmap and map all at once for speed.
    unmap_cmd = [
        "xdotool",
    ]
    map_cmd = [
        "xdotool",
    ]
    t = layout.get_json(window_count)
    if isinstance(t, tuple):
        layout_dict, remap_order = t
        if set(range(window_count)) != set(remap_order):
            raise ValueError("Layout returned invalid remap order")
        windows = [windows[ii] for ii in remap_order]
    else:
        layout_dict = t

    if dry_run:
        print(json.dumps(layout_dict, indent=4))
    else:
        if layout_dict is not False:
            append_layout(layout_dict, window_count)
            for window_id in windows:
                unmap_cmd.append("windowunmap")
                map_cmd.append("windowmap")
                unmap_cmd.append(str(window_id))
                map_cmd.append(str(window_id))

            # force i3 to swallow these windows.
            subprocess.check_call(unmap_cmd)
            subprocess.check_call(map_cmd)
            focus_window(active)


def load_usage():
    try:
        with open(counter_file, "r") as op:
            return json.load(op)
    except (OSError, ValueError):
        return {}


def count_usage(layout_name):
    usage = load_usage()
    if layout_name not in usage:
        usage[layout_name] = (0, datetime.datetime.now().timestamp())
    usage[layout_name] = (
        usage[layout_name][0] + 1,
        datetime.datetime.now().timestamp(),
    )

    with open(counter_file, "w") as op:
        json.dump(usage, op)


def list_layouts_in_smart_order():
    """List the layouts in a 'smart' order,
    that means most common ones on top (by log10 usage),
    within one log10 unit, sorted by most-recently-used"""
    usage = load_usage()
    sort_me = []
    for layout in layouts.layouts:
        if " " in layout.name:
            raise ValueError(
                f"No spaces in layout names please. Offender: '{layout.name}'"
            )
        for alias in [layout.name] + layout.aliases:
            usage_count, last_used = usage.get(
                alias, (0, datetime.datetime.now().timestamp())
            )
            if alias == layout.name:
                desc = alias
            else:
                desc = f"{alias} ({layout.name})"
            sort_me.append(
                (-1 * math.ceil(math.log10(usage_count + 1)), -1 * last_used, desc)
            )
    sort_me.sort()
    for _, _, name in sort_me:
        print(name)


def print_help():
    print(
        """i3-instant-layout applies ready made layouts to i3 workspaces,
    based on the numerical position of the windows.

    Call with '--list' to get a list of available layouts (and their aliases).

    Call with --desc to get detailed information about every layout available.

    Call with the name of a layout to apply it to the current workspace.

    Call with '-' to read layout name from stdin.

    Call with 'name --dry-run' to inspect the generated i3 append_layout compatible json.

    To integrate into i3, add this to your i3/config.
        bindsym $mod+Escape exec "i3-instant-layout --list | rofi -dmenu -i | i3-instant-layout -"

    """
    )
    sys.exit(0)


def print_desc():
    import textwrap

    for layout_class in layouts.layouts:
        print(f"Layout: {layout_class.name}")
        print(f"Aliases: {layout_class.aliases}")
        print(textwrap.indent(textwrap.dedent(layout_class.description), "\t"))
        print("")
        print("-" * 80)
        print("")


def main():
    if len(sys.argv) == 1 or sys.argv[1] == "--help":
        print_help()
    elif sys.argv[1] == "--desc":
        print_desc()
        sys.exit(0)
    elif sys.argv[1] == "--list":
        list_layouts_in_smart_order()
        sys.exit(0)
    elif sys.argv[1] == "-":
        query = sys.stdin.readline().strip()
        print(f'query "{query}"')
        if not query.strip():  # e.g. rofi cancel
            sys.exit(0)
    else:
        query = sys.argv[1]
    if " " in query:
        query = query[: query.find(" ")]
    for layout_class in layouts.layouts:
        if query == layout_class.name or query in layout_class.aliases:
            nuke_swallow_windows()
            apply_layout(layout_class(), "--dry-run" in sys.argv)
            count_usage(query)
            sys.exit(0)
    else:
        print("Could not find the requested layout")
        sys.exit(1)


src/i3_instant_layout/layouts.py
import math

layouts = []


def register_layout(cls):
    layouts.append(cls)
    return cls


def node(percent, layout, swallows, children):
    result = {
        "border": "normal",
        # "current_border_width": 2,
        "floating": "auto_off",
        # "name": "fish  <</home/finkernagel>>",
        "percent": percent,
        "type": "con",
        "layout": layout,
    }
    if swallows:
        result["swallows"] = ([{"class": "."}],)
    if children:
        result["nodes"] = children
    return result


def get_stack(window_count, split):
    return get_stack_unequal([1.0 / window_count] * window_count, split)


def get_stack_unequal(percentages, split):
    elements = []
    for p in percentages:
        elements.append(node(p, split, True, False))
    return [{"layout": split, "type": "con", "nodes": elements}]


@register_layout
class Layout_vStack:

    name = "vStack"
    aliases = ["1col", "1c"]
    description = """\
            One column / a vertical stack.

    ---------
    |   1   |
    ---------
    |   2   |
    ---------
    |   3   |
    ---------
    """

    def get_json(self, window_count):
        return get_stack(window_count, "splitv")


@register_layout
class Layout_hStack:

    name = "hStack"
    aliases = ["1row", "1r"]
    description = """\
            One row / a horizontal stack

    -------------
    |   |   |   |
    | 1 | 2 | 3 |
    |   |   |   |
    -------------
    """

    def get_json(self, window_count):
        return get_stack(window_count, "splith")


@register_layout
class Layout_v2Stack:

    name = "v2Stack"
    aliases = ["2col", "2c", "2v"]
    description = """\
            Two columns of stacks
    -------------
    |  1  |  4  |
    -------------
    |  2  |  5  |
    -------------
    |  3  |  6  |
    -------------
    """

    def get_json(self, window_count):
        s = int(math.ceil(window_count / 2))
        left = get_stack(s, "splitv")
        right = get_stack(s if window_count % 2 == 0 else s - 1, "splitv")
        return [{"layout": "splith", "type": "con", "nodes": [left, right]}]


@register_layout
class Layout_h2Stack:

    name = "h2Stack"
    aliases = ["2row", "2r", "2h"]
    description = """\
            Two rows of stacks
    -------------------
    |  1  |  2  |  3  |
    -------------------
    |  4  |  5  |  6  |
    -------------------
    """

    def get_json(self, window_count):
        s = int(math.ceil(window_count / 2))
        left = get_stack(s, "splith")
        right = get_stack(s if window_count % 2 == 0 else s - 1, "splith")
        return [{"layout": "splitv", "type": "con", "nodes": [left, right]}]


@register_layout
class Layout_v3Stack:

    name = "v3Stack"
    aliases = ["3col", "3c", "3v"]
    description = """\
            Three columns of stacks
    -------------------
    |  1  |  3  |  5  |
    -------------------
    |  2  |  4  |  6  |
    -------------------
    """

    def get_json(self, window_count):
        s = window_count // 3
        a = get_stack(s + window_count % 3, "splitv")
        b = get_stack(s, "splitv")
        c = get_stack(s, "splitv")
        return [{"layout": "splith", "type": "con", "nodes": [a, b, c]}]


@register_layout
class Layout_h3Stack:

    name = "h3Stack"
    aliases = ["3row", "3r", "3h"]
    description = """\
            Three rows of stacks
    -------------------
    |  1  |  2  |  3  |
    -------------------
    |  4  |  5  |  6  |
    -------------------
    |  7  |  8  |  9  |
    -------------------
    """

    def get_json(self, window_count):
        s = window_count // 3
        a = get_stack(s + window_count % 3, "splith")
        b = get_stack(s, "splith")
        c = get_stack(s, "splith")
        return [{"layout": "splitv", "type": "con", "nodes": [a, b, c]}]


@register_layout
class Layout_Max:

    name = "max"
    aliases = ["maxTabbed"]
    description = """\
            One large container,
        in tabbed mode.

       ---------------
       |             |
       |   1,2,3,4,  |
       |             |
       ---------------
    """

    def get_json(self, window_count):
        return get_stack(window_count, "tabbed")


@register_layout
class Layout_MainLeft:

    name = "mainLeft"
    aliases = ["ml", "mv", "MonadTall"]
    description = """\
            One large window to the left at 50%,
            all others stacked to the right vertically.

            -------------
            |     |  2  |
            |     |-----|
            |  1  |  3  |
            |     |-----|
            |     |  4  |
            -------------
            """

    def get_json(self, window_count):
        return node(
            1,
            "splith",
            False,
            [node(0.5, "splitv", True, []), get_stack(window_count - 1, "splitv")],
        )


@register_layout
class Layout_MainRight:

    name = "mainRight"
    aliases = ["mr", "vm", "MonadTallFlip"]
    description = """\
            One large window to the right at 50%,
            all others stacked to the right vertically.

            -------------
            |  2  |     |
            |-----|     |
            |  3  |  1  |
            |-----|     |
            |  4  |     |
            -------------
            """

    def get_json(self, window_count):
        return (
            node(
                1,
                "splith",
                False,
                [get_stack(window_count - 1, "splitv"), node(0.5, "splitv", True, [])],
            ),
            list(range(1, window_count)) + [0],
        )


@register_layout
class Layout_MainMainVStack:

    name = "MainMainVStack"
    aliases = ["mmv"]
    description = """\
            Two large windows to the left at 30%,
            all others stacked to the right vertically.

            -------------------
            |     |     |  3  |
            |     |     |-----|
            |  1  |  2  |  4  |
            |     |     |-----|
            |     |     |  5  |
            -------------------
            """

    def get_json(self, window_count):
        return node(
            1,
            "splith",
            False,
            [
                node(1 / 3, "splitv", True, []),
                node(1 / 3, "splitv", True, []),
                get_stack(window_count - 2, "splitv"),
            ],
        )


@register_layout
class Layout_MainVStackMain:

    name = "MainVStackMain"
    aliases = ["mvm"]
    description = """\
            Two large windows at 30% to the left and right,
            a vstack in the center

            -------------------
            |     |  3  |     |
            |     |-----|     |
            |  1  |  4  |  2  |
            |     |-----|     |
            |     |  5  |     |
            -------------------
            """

    def get_json(self, window_count):
        return (
            node(
                1,
                "splith",
                False,
                [
                    node(1 / 3, "splitv", True, []),
                    get_stack(window_count - 2, "splitv"),
                    node(1 / 3, "splitv", True, []),
                ],
            ),
            [0] + list(range(2, window_count)) + [1],
        )


@register_layout
class Layout_Matrix:

    name = "matrix"
    aliases = []
    description = """\
            Place windows in a n * n matrix.

            The matrix will place swallow-markers
            if you have less than n*n windows.

            N is math.ceil(math.sqrt(window_count))
        """

    def get_json(self, window_count):
        n = int(math.ceil(math.sqrt(window_count)))
        stacks = [get_stack(n, "splith") for stack in range(n)]
        return node(1, "splitv", False, stacks)


@register_layout
class Layout_VerticalTileTop:

    name = "VerticalTileTop"
    aliases = ["vtt"]
    description = """\
            Large master area (66%) on top,
            horizontal stacking below
            """

    def get_json(self, window_count):
        return node(
            1,
            "splitv",
            False,
            [
                node(0.66, "splitv", True, []),
                node(
                    0.33,
                    "splitv",
                    False,
                    get_stack_unequal(
                        [0.33 / (window_count - 1)] * (window_count - 1), "splitv",
                    ),
                ),
            ],
        )


@register_layout
class Layout_VerticalTileBottom:

    name = "VerticalTileBottom"
    aliases = ["vtb"]
    description = """\
            Large master area (66%) on bottom,
            horizontal stacking above
            """

    def get_json(self, window_count):
        return (
            node(
                1,
                "splitv",
                False,
                [
                    node(
                        0.33,
                        "splitv",
                        False,
                        get_stack_unequal(
                            [0.33 / (window_count - 1)] * (window_count - 1), "splitv",
                        ),
                    ),
                    node(0.66, "splitv", True, []),
                ],
            ),
            list(range(1, window_count)) + [0],
        )


@register_layout
class Nested:
    name = "NestedRight"
    aliases = ["nr"]

    description = """\
            Nested layout, starting with a full left half.


            -------------------------
            |           |           |
            |           |     2     |
            |           |           |
            |     1     |-----------|
            |           |     |  4  |
            |           |  3  |-----|
            |           |     |5 | 6|
            -------------------------
            """

    def get_json(self, window_count):
        dir = "h"
        parent = node(1, "splith", False, [])
        root = parent
        parent["nodes"] = []
        for ii in range(window_count):
            parent["nodes"].append(get_stack_unequal([0.5], "split" + dir))
            n = node(1, "splith", False, [])
            if dir == "h":
                dir = "v"
            else:
                dir = "h"

            n["layout"] = "split" + dir
            n["nodes"] = []
            if ii < window_count - 1:
                parent["nodes"].append(n)
                parent = n
        return root


@register_layout
class Smart:
    name = "SmartNestedRight"
    aliases = ["snr"]

    description = """\
            Nested layout, starting with a full left half,
            but never going below 1/16th of the size.

            2 windows
            -------------------------
            |           |           |
            |           |           |
            |           |           |
            |     1     |     2     |
            |           |           |
            |           |           |
            |           |           |
            -------------------------

            5 windows
            -------------------------
            |           |           |
            |           |     2     |
            |           |           |
            |     1     |-----------|
            |           |     |  4  |
            |           |  3  |-----|
            |           |     |  5  |
            -------------------------

            6 windows
            -------------------------
            |           |           |
            |           |     2     |
            |           |           |
            |     1     |-----------|
            |           |  3  |  4  |
            |           |-----|-----|
            |           |  5  |  6  |
            -------------------------

            7 windows
            -------------------------
            |           |     |     |
            |           |  2  |  3  |
            |           |     |     |
            |     1     |-----------|
            |           |  4  |  5  |
            |           |-----|-----|
            |           |  6  |  7  |
            -------------------------


            15 windows
            -------------------------
            |     |  2  |  4  |  6  |
            |  1  |-----|-----|-----|
            |     |  3  |  5  |  7  |
            |-----------|-----------|
            |  8  |  A  |  C  |  E  |
            |-----|-----|-----|-----|
            |  9  |  B  |  D  |  F  |
            -------------------------

            Falls back to matrix layout above 16 windows.
            """

    def get_json(self, window_count):
        def nest_1():
            return node(1, "splith", True, [])

        def nest_2():
            return get_stack(2, "splith")

        def nest_3():
            return node(
                1,
                "splith",
                False,
                [node(0.5, "splitv", True, []), get_stack(2, "splitv")],
            )

        def nest_4():
            return node(
                1, "splith", False, [get_stack(2, "splitv"), get_stack(2, "splitv")],
            )

        if window_count == 1:
            return nest_1()
        elif window_count == 2:
            return nest_2()
        elif window_count == 3:
            return nest_3()
        elif window_count == 4:
            return nest_4()
        elif window_count == 5:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", True, [],),
                    node(
                        0.5, "splitv", False, [node(0.5, "split", True, []), nest_3()]
                    ),
                ],
            )
        elif window_count == 6:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", True, [],),
                    node(
                        0.5, "splitv", False, [node(0.5, "split", True, []), nest_4()]
                    ),
                ],
            )
        elif window_count == 7:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", True, [],),
                    node(0.5, "splitv", False, [nest_2(), nest_4()]),
                ],
            )
        elif window_count == 8:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", True, [],),
                    node(0.5, "splitv", False, [nest_3(), nest_4()]),
                ],
            )
        elif window_count == 9:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", True, [],),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 10:
            return node(
                1,
                "splith",
                False,
                [
                    node(
                        0.5,
                        "splitv",
                        False,
                        [node(0.5, "splitv", True, []), node(0.5, "splitv", True, [])],
                    ),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 11:
            return node(
                1,
                "splith",
                False,
                [
                    node(
                        0.5, "splitv", False, [node(0.5, "splitv", True, []), nest_2()],
                    ),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 12:
            return node(
                1,
                "splith",
                False,
                [
                    node(
                        0.5, "splitv", False, [node(0.5, "splitv", True, []), nest_3()],
                    ),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 13:
            return node(
                1,
                "splith",
                False,
                [
                    node(
                        0.5, "splitv", False, [node(0.5, "splitv", True, []), nest_4()],
                    ),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 14:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", False, [nest_2(), nest_4()],),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 15:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", False, [nest_3(), nest_4()],),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        elif window_count == 16:
            return node(
                1,
                "splith",
                False,
                [
                    node(0.5, "splitv", False, [nest_4(), nest_4()],),
                    node(0.5, "splitv", False, [nest_4(), nest_4()]),
                ],
            )
        else:
            return Layout_Matrix().get_json(window_count)

@register_layout
class Layout_MainCenter:

    name = "mainCenter"
    aliases = ["mc", "vmv"]
    description = """\
            One large window in the midle at 50%,
            all others stacked to the left/right vertically.

            -------------------
            |  2  |     |  5  |
            |-----|     |-----|
            |  3  |  1  |  6  |
            |-----|     |-----|
            |  4  |     |  7  |
            -------------------
            """

    def get_json(self, window_count):
        lr = window_count - 1
        left = math.ceil(lr / 2)
        right = math.floor(lr / 2)
        nodes = []
        if left:
            nodes.append(node(0.25, 'splith', False, get_stack(left, 'splitv')))
        nodes.append(node(0.5, 'splitv', True, []))
        if right:
            nodes.append(node(0.25, 'splith', False, get_stack(right, 'splitv')))
        order = list(range(1, left+1)) + [0] + list(range(left+1, left+1+right))
        print(order)
        return node(1, 'splith', False, nodes),order

